controller/bindings/connman/connman.ml

```diff
@@ -82,60 +89,135 @@ struct
+     produce. Defined but unused errors are not included. *)
+  type manager_error =
+      | InvalidKey
+      | ConnectFailed
+      | Blocked
+      | Unknown of string
+
+  type agent_error =
+      | PassMissing
+      | MissingRequestedInputs of string list
+      | BrowserRequested
+      | UnexpectedFailure of exn
+
+  type reported_error =
+      | ManagerError of manager_error
+      | AgentError of agent_error
+
   let request_input input service (fields: (string * OBus_value.V.single) list) =
     let%lwt () = Logs_lwt.debug ~src:log_src
-        (fun m -> m "input requested from agent for service %s"
+        (fun m -> m "input %s requested from agent for service %s"
+            (String.concat ", " @@ List.map (fun (k, v) ->
+                Printf.sprintf "%s: %s" k (OBus_value.V.string_of_single v)
+            ) fields)
             (String.concat "/" service)
         )
     in
-    match input with
-    | Passphrase p ->
-      (match List.assoc_opt "Passphrase" fields with
-       | Some _ ->
-         return [ "Passphrase", p |> OBus_value.C.(make_single basic_string)]
-       | None ->
-         let%lwt () = Logs_lwt.err ~src:log_src
-             (fun m -> m "Passphrase available as input but not being requested")
-         in
-         OBus_error.make "net.connman.Agent.Error.Canceled" "."
-         |> Lwt.fail
-      )
-    | None ->
-      let%lwt () = Logs_lwt.err ~src:log_src
-          (fun m -> m "input requested from agent but none available.")
-      in
-      OBus_error.make "net.connman.Agent.Error.Canceled" "No input available."
-      |> Lwt.fail
-
+    let mandatory_inputs = List.concat_map (fun (k, v) ->
+        let v_ocaml = OBus_value.C.(cast_single (dict string variant) v) in
+        let requirement_opt = List.assoc_opt "Requirement" v_ocaml |>
+            Option.map (OBus_value.C.(cast_single basic_string)) in
+        match requirement_opt with
+            | Some "mandatory" -> [k]
+            | _ -> []
+    ) fields
+    in
+    match (input, mandatory_inputs) with
+        | (Passphrase p, [ "Passphrase" ]) ->
+            return @@ Ok [ "Passphrase", p |> OBus_value.C.(make_single basic_string)]
+        | (None, [ "Passphrase" ]) ->
+          let%lwt () = Logs_lwt.err ~src:log_src (fun m -> m
+            "Passphrase requested from agent, but not available.")
+          in
+          return @@ Error PassMissing
+        | (_, _) ->
+            (* This case is triggered when manager is requesting
+               _some_ input(s) (so NOT open networks) and those inputs
+               involve something additional to (or other than) a passphrase.
+            *)
+            let expected_properties_str =
+                String.concat ", " mandatory_inputs
+            in
+            let%lwt () = Logs_lwt.err ~src:log_src (fun m -> m
+               "Manager is requesting additional missing input(s): %s"
+               expected_properties_str
```

Comment:
Nitpick: This would seem to imply that the password is missing, even when it has been entered. It's also not so clear to a potential users who the "Manager" is.

Rephrase to state that the _network_ is requesting missing inputs, and give the full list of inputs? (Alternative: Filter out passphrase from list if one was given.)

`"This network requires additional inputs for authentication. The network requests the following: %s"`

Reply:
Of course, I was in read-string-translations mode and completely ignored the surrounding lines ...